impl/spanner/mutations/table.go

```diff
@@ -0,0 +1,311 @@
+		func(row *spanner.Row) error {
+			var logID int64
+			if err := row.Columns(&logID); err != nil {
+				return err
+			}
+			logIDs = append(logIDs, logID)
+			return nil
+		}); err != nil {
+		return nil, err
+	}
+	if len(logIDs) == 0 {
+		return nil, status.Errorf(codes.NotFound, "not found")
+	}
+	return logIDs, nil
+}
+
+// Send submits an item to the queue and returns the commit timestamp.
+func (t *Table) Send(ctx context.Context, directoryID string, logID int64, entries ...*pb.EntryUpdate) (water.Mark, error) {
+	if len(entries) == 0 {
+		return water.Mark{}, nil
+	}
+	type Cols struct {
+		DirectoryID string
+		LogID       int64
+		Timestamp   time.Time
+		LocalID     int64
+		Mutation    []byte
+	}
+	ms := make([]*spanner.Mutation, 0, len(entries))
+	for i, e := range entries {
+		mBytes, err := proto.Marshal(e)
+		if err != nil {
+			return water.Mark{}, err
+		}
+		m, err := spanner.InsertStruct(mutTable, Cols{
+			DirectoryID: directoryID,
+			LogID:       logID,
+			Timestamp:   spanner.CommitTimestamp,
+			LocalID:     int64(i),
+			Mutation:    mBytes,
+		})
+		if err != nil {
+			return water.Mark{}, err
+		}
+		ms = append(ms, m)
+	}
+
+	commitTimestamp, err := t.client.ReadWriteTransaction(ctx,
+		func(_ context.Context, txn *spanner.ReadWriteTransaction) error {
+			return txn.BufferWrite(ms)
+		})
+	return timeToMark(commitTimestamp), err
+}
+
+// HighWatermark returns the number of items and the highest primary
+// key (exclusive) up to batchSize items after start (inclusive).
+func (t *Table) HighWatermark(ctx context.Context, directoryID string, logID int64, start water.Mark, batchSize int32) (count int32, high water.Mark, err error) {
+	// TODO: Replace with MAX(Revision) when spansql supports aggregate operators.
+	stmt := spanner.NewStatement(`
+	SELECT Timestamp FROM Mutations WHERE
+		DirectoryID = @directoryID AND
+		LogID = @logID AND
+		Timestamp >= @start
+	ORDER BY Timestamp
+	LIMIT @limit`)
+	stmt.Params["directoryID"] = directoryID
+	stmt.Params["logID"] = logID
+	stmt.Params["start"] = markToTime(start)
+	stmt.Params["limit"] = int64(batchSize)
+
+	var cnt int64
+	var max time.Time
+	rtx := t.client.Single().WithTimestampBound(spanner.MaxStaleness(readStaleness))
+	defer rtx.Close()
+	if err := rtx.Query(ctx, stmt).Do(
+		func(row *spanner.Row) error {
+			cnt++
+			var watermark time.Time
+			if err := row.Column(0, &watermark); err != nil {
+				return err
+			}
+			if watermark.After(max) {
+				max = watermark
+			}
+			return nil
+		}); err != nil {
+		return 0, water.Mark{}, err
+	}
+	if cnt == 0 {
+		// When there are no rows, return the start time as the highest timestamp.
+		return 0, start, nil
+	}
+	return int32(cnt), timeToMark(max).Add(1), nil
+}
+
+// ReadLog reads all mutations between [low, high).
+func (t *Table) ReadLog(ctx context.Context, directoryID string, logID int64, low, high water.Mark,
+	limit int32) ([]*mutator.LogMessage, error) {
+	if high.Value() == 0 || limit == 0 {
```

Comment:
Similar to above, I worry that this would be an application bug and should fail fast instead of limping on. But this is a decision that should be consistent across the library and you're in the right position to judge that.

Reply:
I suppose I'm coming from a maths perspective where we define things for edge cases like "1 is a prime"

Reply:
Limit = 0 and high =0 are valid values to pass in, even though they might indicative of logic errors higher in the stack. Because this function doesn't "know" for sure that they are logic errors, they should not be treated as such here. Happy to discuss this